airflow/models/renderedtifields.py

```diff
@@ -0,0 +1,135 @@
+from sqlalchemy import Column, String, and_, tuple_
+from sqlalchemy.orm import Session
+
+from airflow.configuration import conf
+from airflow.models.base import ID_LEN, Base
+from airflow.models.taskinstance import TaskInstance
+from airflow.serialization.helpers import serialize_template_field
+from airflow.settings import json
+from airflow.utils.session import provide_session
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+class RenderedTaskInstanceFields(Base):
+    """
+    Save Rendered Template Fields
+    """
+
+    __tablename__ = "rendered_task_instance_fields"
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    task_id = Column(String(ID_LEN), primary_key=True)
+    execution_date = Column(UtcDateTime, primary_key=True)
+    rendered_fields = Column(sqlalchemy_jsonfield.JSONField(json=json), nullable=False)
+
+    def __init__(self, ti: TaskInstance, render_templates=True):
+        self.dag_id = ti.dag_id
+        self.task_id = ti.task_id
+        self.task = ti.task
+        self.execution_date = ti.execution_date
+        self.ti = ti
+        if render_templates:
+            ti.render_templates()
+        self.rendered_fields = {
+            field: serialize_template_field(
+                getattr(self.task, field)
+            ) for field in self.task.template_fields
+        }
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.dag_id}.{self.task_id} {self.execution_date}"
+
+    @classmethod
+    @provide_session
+    def get_templated_fields(cls, ti: TaskInstance, session: Session = None) -> Optional[dict]:
+        """
+        Get templated field for a TaskInstance from the RenderedTaskInstanceFields
+        table.
+
+        :param ti: Task Instance
+        :param session: SqlAlchemy Session
+        :return: Rendered Templated TI field
+        """
+        result = session.query(cls.rendered_fields).filter(
+            cls.dag_id == ti.dag_id,
+            cls.task_id == ti.task_id,
+            cls.execution_date == ti.execution_date
+        ).one_or_none()
+
+        if result:
+            rendered_fields = result.rendered_fields
+            return rendered_fields
+        else:
+            return None
+
+    @provide_session
+    def write(self, session: Session = None):
+        """Write instance to database
+
+        :param session: SqlAlchemy Session
+        """
+        session.merge(self)
+
+    @classmethod
+    @provide_session
+    def delete_old_records(
+        cls, task_id: str, dag_id: str,
+        num_to_keep=conf.getint("core", "max_num_rendered_ti_fields_per_task", fallback=0),
+        session: Session = None
+    ):
+        """
+        Keep only Last X (num_to_keep) number of records for a task by deleting others
+
+        :param task_id: Task ID
+        :param dag_id: Dag ID
+        :param num_to_keep: Number of Records to keep
+        :param session: SqlAlchemy Session
+        """
+        if num_to_keep > 0:
```

Comment:
Can you reverse the condition to reduce the number of indentations in the code?

Reply:
I like to leave the function as soon as possible. This way, when I analyze the code, I don't have to read the whole function, but only the key fragment. This also limits the appearance of too-many-nested-blocks error in pylint.

Reply:
Updated